gpMgmt/bin/gpcheckcat_modules/leaked_schema_dropper.py

```diff
@@ -3,35 +3,36 @@
 
     # This query does a union of all the leaked temp schemas on the coordinator as well as all the segments.
-    # The first part of the query uses gp_dist_random which gets the leaked schemas from only the segments
-    # The second part of the query gets the leaked temp schemas from just the coordinator
-    # The simpler form of this query that pushed the union into the
-    # inner select does not run correctly on 3.2.x
     # Also note autovacuum lancher and worker will not generate temp namespace
+    # Additionally, we should take into account orphaned schemas, whose session_id may
+    # conflict with coordinator bgworkers, which shouldn't create orphaned schemas.
+    # Also try to avoid connection id collision with truly orphaned schema id.
     leaked_schema_query = """
-        SELECT distinct nspname as schema
-        FROM (
-          SELECT nspname, replace(nspname, 'pg_temp_','')::int as sess_id
-          FROM   gp_dist_random('pg_namespace')
-          WHERE  nspname ~ '^pg_temp_[0-9]+'
-          UNION ALL
-          SELECT nspname, replace(nspname, 'pg_toast_temp_','')::int as sess_id
-          FROM   gp_dist_random('pg_namespace')
-          WHERE  nspname ~ '^pg_toast_temp_[0-9]+'
-        ) n LEFT OUTER JOIN pg_stat_activity x using (sess_id)
-        WHERE x.sess_id is null OR x.backend_type like 'autovacuum%'
-        UNION
-        SELECT nspname as schema
-        FROM (
-          SELECT nspname, replace(nspname, 'pg_temp_','')::int as sess_id
-          FROM   pg_namespace
-          WHERE  nspname ~ '^pg_temp_[0-9]+'
-          UNION ALL
-          SELECT nspname, replace(nspname, 'pg_toast_temp_','')::int as sess_id
-          FROM   pg_namespace
-          WHERE  nspname ~ '^pg_toast_temp_[0-9]+'
-        ) n LEFT OUTER JOIN pg_stat_activity x using (sess_id)
-        WHERE x.sess_id is null OR x.backend_type like 'autovacuum%'
+        WITH temp_schemas AS (
+          SELECT nspname, 
+                 CASE 
+                   WHEN nspname ~ '^pg_temp_[0-9]+' THEN replace(nspname, 'pg_temp_', '')::int
+                   ELSE replace(nspname, 'pg_toast_temp_', '')::int
+                 END as sess_id
+          FROM gp_dist_random('pg_namespace')
+          WHERE nspname ~ '^pg_t(emp|oast_temp)_[0-9]+'
+          UNION
+          SELECT nspname,
+                 CASE 
+                   WHEN nspname ~ '^pg_temp_[0-9]+' THEN replace(nspname, 'pg_temp_', '')::int
+                   ELSE replace(nspname, 'pg_toast_temp_', '')::int
+                 END as sess_id
+          FROM pg_namespace
+          WHERE nspname ~ '^pg_t(emp|oast_temp)_[0-9]+'
+        )
+        SELECT DISTINCT nspname as schema
+        FROM temp_schemas n 
+        LEFT JOIN pg_stat_activity x USING (sess_id)
+        WHERE x.sess_id IS NULL 
+           OR x.backend_type LIKE 'autovacuum%'
+           OR x.datname IS NULL 
+           OR x.datname <> current_database()
+           OR x.pid = pg_backend_pid()
```

Comment:
Could you please elaborate on the meaning of these three new additional conditions?

Reply:
> 1. x.datname <> current_database()

We ran `gpcheckcat` in one database, but we want it to delete schemas in another? How will it do this if it's not connected to it?